File: packages/glob/src/format.ts
/**
 * Context the zero-match hint needs to tailor its suggestions.
 */
export interface ZeroMatchContext {
  readonly hasRecursiveMarker: boolean;
  readonly explicitPath: boolean;
}

function zeroMatchHint(pattern: string, ctx: ZeroMatchContext): string {
  const suggestions: string[] = [];
  if (!ctx.hasRecursiveMarker) {
    suggestions.push(
      `add '**/' before the pattern to search recursively (e.g. '**/${pattern}')`,
    );
  }
  suggestions.push(
    "broaden the pattern (e.g. replace '.ts' with '.{ts,tsx,js}')",
  );
  if (ctx.explicitPath) {
    suggestions.push("try a different path, or omit 'path' to search the workspace root");
  } else {
    suggestions.push("try a different path");
  }
  return `(No files matched '${pattern}'. Try: ${suggestions.join("; ")}.)`;
}

export function hasRecursiveMarker(pattern: string): boolean {
  return /\*\*/.test(pattern);
}

/**
 * Suggestions for narrowing an over-broad pattern. Kept as concrete
 * templated examples because weak models pattern-match on the example
 * rather than the abstract rule. The order matters: scope → extension
 * → anchor is the same decision tree a human would walk.
 *
 * explicitPath lets us surface the anchor option only when `path` was
 * defaulted — otherwise we're suggesting something the model already did.
 */
function narrowingSuggestions(pattern: string, explicitPath: boolean): string[] {
  const hasExt = /\.[a-zA-Z0-9]+(?:\}|$)/.test(pattern);
  const out: string[] = [];
  out.push(
    "scope to a subdirectory (e.g. 'src/" +
      (pattern.startsWith("**/") ? pattern.slice(3) : pattern) +
      "')",
  );
  if (!hasExt) {
    out.push("pick a specific file extension (e.g. '**/*.ts' or '**/*.md')");
  } else {
    out.push("tighten the extension set");
  }
  if (!explicitPath) {
    out.push(
      "use the 'path' parameter to anchor the search in a subdirectory",
    );
  }
  return out;
}

export function formatPaths(params: {
  pattern: string;
  paths: readonly string[];
  total: number;
  offset: number;
  headLimit: number;
  more: boolean;
  zeroMatchContext: ZeroMatchContext;
}): string {
  const { pattern, paths, total, offset, more, zeroMatchContext } = params;
  const header = `<pattern>${pattern}</pattern>\n<paths>`;
  if (paths.length === 0) {
    return `${header}\n${zeroMatchHint(pattern, zeroMatchContext)}\n</paths>`;
  }
  const body = paths.join("\n");
  const next = offset + paths.length;
  let hint: string;
  if (!more) {
    hint = `(Found ${total} file(s) matching the pattern.)`;
  } else {
    // Truncated result: lead with narrowing, demote pagination to fallback.
    // Heuristic "likely broader than intended":
    // - total is >= 4× the headLimit (runaway match count), OR
    // - pattern is a bare catch-all: '*', '**', '**/*', '**/**' — these
    //   match anything at any depth and almost always indicate the model
    //   didn't think about what it actually wanted.
    // A specific-extension pattern like '**/*.ts' is NOT broad by this
    // heuristic; the model can still page or narrow by subdirectory.
    const bareCatchAll = /^(\*|\*\*|\*\*\/\*|\*\*\/\*\*)$/.test(pattern);
    const isVeryBroad = bareCatchAll || total >= params.headLimit * 4;
    const narrow = narrowingSuggestions(pattern, zeroMatchContext.explicitPath);
    const showing = `(Showing files ${offset + 1}-${next} of ${total} matching '${pattern}'.`;
    const broadNote = isVeryBroad
      ? " This is likely broader than intended."
      : "";
    const narrowLine = `\nTo narrow: ${narrow.join("; ")}.`;
    const pageLine = `\nTo page through instead, re-call with offset: ${next}.)`;
    hint = showing + broadNote + narrowLine + pageLine;
  }
  return `${header}\n${body}\n\n${hint}\n</paths>`;
}

File: crates/glob/src/format.rs
pub fn has_recursive_marker(pattern: &str) -> bool {
    pattern.contains("**")
}

#[derive(Debug, Clone, Copy)]
pub struct ZeroMatchContext {
    pub has_recursive_marker: bool,
    pub explicit_path: bool,
}

fn zero_match_hint(pattern: &str, ctx: ZeroMatchContext) -> String {
    let mut suggestions: Vec<String> = Vec::new();
    if !ctx.has_recursive_marker {
        suggestions.push(format!(
            "add '**/' before the pattern to search recursively (e.g. '**/{}')",
            pattern
        ));
    }
    suggestions.push("broaden the pattern (e.g. replace '.ts' with '.{ts,tsx,js}')".to_string());
    if ctx.explicit_path {
        suggestions.push("try a different path, or omit 'path' to search the workspace root".to_string());
    } else {
        suggestions.push("try a different path".to_string());
    }
    format!("(No files matched '{}'. Try: {}.)", pattern, suggestions.join("; "))
}

/// Mirror the `narrowingSuggestions` helper from the TS side.
fn narrowing_suggestions(pattern: &str, explicit_path: bool) -> Vec<String> {
    let has_ext = regex_like_has_ext(pattern);
    let mut out: Vec<String> = Vec::new();
    let tail = if let Some(rest) = pattern.strip_prefix("**/") {
        rest.to_string()
    } else {
        pattern.to_string()
    };
    out.push(format!("scope to a subdirectory (e.g. 'src/{}')", tail));
    if !has_ext {
        out.push("pick a specific file extension (e.g. '**/*.ts' or '**/*.md')".to_string());
    } else {
        out.push("tighten the extension set".to_string());
    }
    if !explicit_path {
        out.push("use the 'path' parameter to anchor the search in a subdirectory".to_string());
    }
    out
}

fn regex_like_has_ext(pattern: &str) -> bool {
    // Roughly matches `/\.[a-zA-Z0-9]+(?:\}|$)/` — "has an extension
    // token near the tail". Cheap heuristic.
    let trimmed = pattern.trim_end_matches('}');
    match trimmed.rfind('.') {
        None => false,
        Some(i) => {
            if i + 1 >= pattern.len() {
                return false;
            }
            let tail = &pattern[i + 1..];
            tail.chars().all(|c| c.is_ascii_alphanumeric() || c == '}' || c == ',' || c == '{')
        }
    }
}

pub struct FormatPathsArgs<'a> {
    pub pattern: &'a str,
    pub paths: &'a [String],
    pub total: usize,
    pub offset: usize,
    pub head_limit: usize,
    pub more: bool,
    pub zero_match_context: ZeroMatchContext,
}

pub fn format_paths(args: FormatPathsArgs<'_>) -> String {
    let header = format!("<pattern>{}</pattern>\n<paths>", args.pattern);
    if args.paths.is_empty() {
        let hint = zero_match_hint(args.pattern, args.zero_match_context);
        return format!("{}\n{}\n</paths>", header, hint);
    }
    let body = args.paths.join("\n");
    let next = args.offset + args.paths.len();
    let hint = if !args.more {
        format!("(Found {} file(s) matching the pattern.)", args.total)
    } else {
        // Truncated: lead with narrowing, demote pagination.
        let bare_catch_all = matches!(args.pattern, "*" | "**" | "**/*" | "**/**");
        let is_very_broad = bare_catch_all || args.total >= args.head_limit * 4;
        let narrow = narrowing_suggestions(args.pattern, args.zero_match_context.explicit_path);
        let showing = format!(
            "(Showing files {}-{} of {} matching '{}'.",
            args.offset + 1,
            next,
            args.total,
            args.pattern
        );
        let broad = if is_very_broad {
            " This is likely broader than intended."
        } else {
            ""
        };
        let narrow_line = format!("\nTo narrow: {}.", narrow.join("; "));
        let page_line = format!("\nTo page through instead, re-call with offset: {}.)", next);
        format!("{}{}{}{}", showing, broad, narrow_line, page_line)
    };
    format!("{}\n{}\n\n{}\n</paths>", header, body, hint)
}
